test/testwave/testfiles/t_8_001.cpp

```diff
@@ -0,0 +1,44 @@
+/*=============================================================================
+    Boost.Wave: A Standard compliant C++ preprocessor library
+    http://www.boost.org/
+
+    Copyright (c) 2020 Jeff Trull. Distributed under the Boost
+    Software License, Version 1.0. (See accompanying file
+    LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+=============================================================================*/
+
+//O --c++20
+//O -Werror
+
+// __VA_OPT__ examples from the proposal doc P00306
+// Those examples have a few weird typos. I've tried to match the
+// behavior of gcc in the expected results - JET
+
+#define LOG(msg, ...) printf(msg __VA_OPT__(,) __VA_ARGS__)
+#define SDEF(sname, ...) S sname __VA_OPT__(= { __VA_ARGS__ })
+#define LOG2(...)                                     \
+    printf("at line=%d" __VA_OPT__(": "), __LINE__);  \
+    __VA_OPT__(printf(__VA_ARGS__);)                  \
+    printf("\n")
+
+//R #line 28 "t_8_001.cpp"
+//R printf("hello world\n" );
+//R printf("hello world\n" );
+//R printf("hello %d\n" , n);
+LOG("hello world\n");
+LOG("hello world\n", );
+LOG("hello %d\n", n);
+
+//R #line 35 "t_8_001.cpp"
+//R S foo;
+//R S bar = { 1, 2, 3 };
+SDEF(foo);
+SDEF(bar, 1, 2, 3);
+
+//R #line 43 "t_8_001.cpp"
+//R printf("at line=%d" , 43); printf("\n");
+// P00306 example suggests the adjacent string literals are concatenated in
+// this test case but that's the *compiler's* job:
+//R printf("at line=%d" ": ", 44); printf("All well in zone %d", n); printf("\n");
+LOG2();
+LOG2("All well in zone %d", n);
```

Comment:
Do you think we should add the hook verification as well (i.e. `//H...` comments)?

Reply:
FWIW, adding a `//H` line at the end of t_8_001.cpp makes testwave produce the following recorded hooks:
```c++
//H 10: t_8_001.cpp(17) : #define
//H 08: t_8_001.cpp(17) : LOG(msg, ...) = printf(msg __VA_OPT__(, ) __VA_ARGS__)
//H 10: t_8_001.cpp(18) : #define
//H 08: t_8_001.cpp(18) : SDEF(sname, ...) = S sname __VA_OPT__(= { __VA_ARGS__ })
//H 10: t_8_001.cpp(19) : #define
//H 08: t_8_001.cpp(19) : LOG2(...) = printf("at line=%d" __VA_OPT__(": "), __LINE__);      __VA_OPT__(printf(__VA_ARGS__);)                      printf("\n")
//H 00: t_8_001.cpp(28) : LOG("hello world\n"), [t_8_001.cpp(17):LOG(msg, ...) = printf(msg __VA_OPT__(, ) __VA_ARGS__)]
//H 02: printf("hello world\n" º)
//H 03: printf("hello world\n")
//H 00: t_8_001.cpp(29) : LOG("hello world\n", º), [t_8_001.cpp(17):LOG(msg, ...) = printf(msg __VA_OPT__(, ) __VA_ARGS__)]
//H 02: printf("hello world\n" º)
//H 03: printf("hello world\n")
//H 00: t_8_001.cpp(30) : LOG("hello %d\n", n), [t_8_001.cpp(17):LOG(msg, ...) = printf(msg __VA_OPT__(, ) __VA_ARGS__)]
//H 02: printf("hello %d\n", n)
//H 03: printf("hello %d\n", n)
//H 00: t_8_001.cpp(35) : SDEF(foo), [t_8_001.cpp(18):SDEF(sname, ...) = S sname __VA_OPT__(= { __VA_ARGS__ })]
//H 02: S foo º
//H 03: S foo
//H 00: t_8_001.cpp(36) : SDEF(bar, 1, 2, 3), [t_8_001.cpp(18):SDEF(sname, ...) = S sname __VA_OPT__(= { __VA_ARGS__ })]
//H 02: S bar = { 1, 2, 3 }
//H 03: S bar = { 1, 2, 3 }
//H 00: t_8_001.cpp(43) : LOG2(º), [t_8_001.cpp(19):LOG2(...) = printf("at line=%d" __VA_OPT__(": "), __LINE__);      __VA_OPT__(printf(__VA_ARGS__);)                      printf("\n")]
//H 02: printf("at line=%d" º, __LINE__);      º                      printf("\n")
//H 03: printf("at line=%d", 43);                            printf("\n")
//H 00: t_8_001.cpp(44) : LOG2("All well in zone %d", n), [t_8_001.cpp(19):LOG2(...) = printf("at line=%d" __VA_OPT__(": "), __LINE__);      __VA_OPT__(printf(__VA_ARGS__);)                      printf("\n")]
//H 02: printf("at line=%d" ": ", __LINE__);      printf("All well in zone %d", n);                      printf("\n")
//H 03: printf("at line=%d" ": ", 44);      printf("All well in zone %d", n);                      printf("\n")
```
This looks bogus as it apparently contains placemarker tokens (the `º` characters). I think that shouldn't happen.

Reply:
Actually, the placemarker symbols may only become removed after the `H 02` (`expanded_macro` hook) is called (during rescanning of the replacement list). Thus, seeing those in the output could be correct after all. I just have never seen any placemarkers being printed in this context before. 

We might want to remove placemarkers from the generated string representation, similarly to what we do while generating string literals (see: https://github.com/boostorg/wave/blob/develop/include/boost/wave/util/macro_helpers.hpp#L128-L135).

Reply:
1. I will add hook verification.
2. I will investigate whether the placemarkers are new, and if not, look at removing them from string reps.

Reply:
It looks like on develop branch `t_1_032.cpp`, which I haven't touched, also shows placemarkers in the output, so I think this is not new. I will investigate scrubbing them.

Reply:
Ok, I wasn't even aware of this anymore... In that case, sorry for the noise. It might be even correct to leave them in to ensure proper functioning (for instance to make sure that we don't see two of them in a row, etc.)

Reply:
I've reviewed the hook output and concluded it's correct, so I've updated the testcase accordingly.